core/lib/zksync_core/src/state_keeper/io/seal_logic/l2_block_seal_subtasks.rs

```diff
@@ -0,0 +1,584 @@
+use zksync_db_connection::{connection::Connection, connection_pool::ConnectionPool};
+use zksync_types::{event::extract_added_tokens, L2BlockNumber};
+
+use crate::state_keeper::{
+    metrics::{L2BlockSealStage, L2_BLOCK_METRICS},
+    updates::L2BlockSealCommand,
+};
+
+/// Helper struct that encapsulates parallel l2 block sealing logic.
+#[derive(Debug)]
+pub(crate) struct L2BlockSealProcess;
+
+impl L2BlockSealProcess {
+    pub fn all_subtasks() -> Vec<Box<dyn L2BlockSealSubtask>> {
+        vec![
+            Box::new(MarkTransactionsInL2BlockSubtask),
+            Box::new(InsertStorageLogsSubtask),
+            Box::new(InsertFactoryDepsSubtask),
+            Box::new(InsertTokensSubtask),
+            Box::new(InsertEventsSubtask),
+            Box::new(InsertL2ToL1LogsSubtask),
+        ]
+    }
+
+    pub fn subtasks_len() -> u32 {
+        Self::all_subtasks().len() as u32
+    }
+
+    pub async fn run_subtasks(
+        command: Arc<L2BlockSealCommand>,
+        pool: ConnectionPool<Core>,
+    ) -> anyhow::Result<()> {
+        let subtasks = Self::all_subtasks();
+        let handles: Vec<_> = subtasks
+            .into_iter()
+            .map(|subtask| {
+                subtask
+                    .spawn(command.clone(), pool.clone())
+                    .map(|res| res.unwrap_or_else(|err| Err(err.into())))
+            })
+            .collect();
+        futures::future::try_join_all(handles).await?;
+
+        Ok(())
+    }
+}
+
+/// An abstraction that represents l2 block seal sub-task that can be run in parallel with other sub-tasks.
+#[async_trait::async_trait]
+pub(crate) trait L2BlockSealSubtask: Send + Sync + 'static {
+    /// Returns sub-task name.
+    fn name(&self) -> &'static str;
+
+    /// Spawns tokio task that runs the seal process.
+    fn spawn(
+        self: Box<Self>,
+        command: Arc<L2BlockSealCommand>,
+        pool: ConnectionPool<Core>,
+    ) -> JoinHandle<anyhow::Result<()>>;
+
+    /// Rollbacks data that was saved to database for the pending L2 block.
+    async fn rollback(
+        &self,
+        storage: &mut Connection<'_, Core>,
+        last_sealed_l2_block: L2BlockNumber,
+    ) -> anyhow::Result<()>;
+}
+
+#[derive(Debug)]
+pub(super) struct MarkTransactionsInL2BlockSubtask;
+
+#[async_trait]
+impl L2BlockSealSubtask for MarkTransactionsInL2BlockSubtask {
+    fn name(&self) -> &'static str {
+        "mark_transactions_in_l2_block"
+    }
+
+    fn spawn(
+        self: Box<Self>,
+        command: Arc<L2BlockSealCommand>,
+        pool: ConnectionPool<Core>,
+    ) -> JoinHandle<anyhow::Result<()>> {
+        tokio::spawn(async move {
+            let progress = L2_BLOCK_METRICS.start(
+                L2BlockSealStage::MarkTransactionsInL2Block,
+                command.is_l2_block_fictive(),
+            );
+
+            let mut connection = pool
+                .connection_tagged("state_keeper")
+                .await
+                .context(self.name())?;
```

Comment:
Bikeshedding: IMO, it'd make more sense to add context to the call site of `spawn()` (or the equivalent `run()` method; see the other comment), so that it's clear that it'll be added to all errors uniformly, and requires less code.